cartographer_ros/src/sensor_bridge.cc

```diff
@@ -16,75 +16,143 @@
-  sensor_collator_->AddSensorData(
-      trajectory_id_,
-      ::cartographer::common::ToUniversal(FromRos(msg->header.stamp)), topic,
-      std::move(sensor_data));
+void SensorBridge::HandleOdometryMessage(
+    const string& topic, const nav_msgs::Odometry::ConstPtr& msg) {
+  const carto::common::Time time = FromRos(msg->header.stamp);
+  const Eigen::Matrix3d translational =
+      Eigen::Matrix3d::Identity() *
+      options_.constant_odometry_translational_variance;
+  const Eigen::Matrix3d rotational =
+      Eigen::Matrix3d::Identity() *
+      options_.constant_odometry_rotational_variance;
+  // clang-format off
+  carto::kalman_filter::PoseCovariance covariance;
+  covariance <<
+      translational, Eigen::Matrix3d::Zero(),
+      Eigen::Matrix3d::Zero(), rotational;
+  // clang-format on
+  const auto sensor_to_tracking =
+      tf_bridge_->LookupToTracking(time, msg->child_frame_id);
+  if (sensor_to_tracking != nullptr) {
+    sensor_collator_->AddSensorData(
+        trajectory_id_, carto::common::ToUniversal(time), topic,
+        carto::common::make_unique<SensorData>(
+            msg->child_frame_id,
+            SensorData::Odometry{
+                ToRigid3d(msg->pose.pose) * sensor_to_tracking->inverse(),
+                covariance}));
+  }
 }
 
-void SensorBridge::AddImuMessage(const string& topic,
-                                 const sensor_msgs::Imu::ConstPtr& msg) {
+void SensorBridge::HandleImuMessage(const string& topic,
+                                    const sensor_msgs::Imu::ConstPtr& msg) {
   CHECK_NE(msg->angular_velocity_covariance[0], -1);
   CHECK_NE(msg->linear_acceleration_covariance[0], -1);
-  auto sensor_data = ::cartographer::common::make_unique<SensorData>(
-      msg->header.frame_id,
-      SensorData::Imu{{msg->angular_velocity.x, msg->angular_velocity.y,
-                       msg->angular_velocity.z},
-                      {msg->linear_acceleration.x, msg->linear_acceleration.y,
-                       msg->linear_acceleration.z}});
-  sensor_collator_->AddSensorData(
-      trajectory_id_,
-      ::cartographer::common::ToUniversal(FromRos(msg->header.stamp)), topic,
-      std::move(sensor_data));
+  const carto::common::Time time = FromRos(msg->header.stamp);
+  const auto sensor_to_tracking =
+      tf_bridge_->LookupToTracking(time, msg->header.frame_id);
+  if (sensor_to_tracking != nullptr) {
+    CHECK(sensor_to_tracking->translation().norm() < 1e-5)
+        << "The IMU frame must be colocated with the tracking frame. "
+           "Transforming linear acceleration into the tracking frame will "
+           "otherwise be imprecise.";
+    sensor_collator_->AddSensorData(
+        trajectory_id_, carto::common::ToUniversal(time), topic,
+        carto::common::make_unique<SensorData>(
+            msg->header.frame_id,
+            SensorData::Imu{
+                sensor_to_tracking->rotation() * ToEigen(msg->angular_velocity),
```

Comment:
Swap angular velocity/linear accelerations for consistency: The variance options, the covariance matrix and the cartographer library all have the opposite ordering.